### DISCUSSION TITLE
Better error messages

### DISCUSSION THREAD

**Discussion by ed2050** [zauberzeug/nicegui#2167] [Ideas / Feature Requests]
Nicegui would benefit from better error messages in certain situations.  Sometimes it's not clear at all what causes an error.

Consider the error traceback below.  The nicegui error doesn't reference my code at all, only libraries.  I would have no idea what triggered this error based on the exception message.

# Error

Here's the error tracback in the console.

```py
'tuple' object has no attribute 'startswith'
Traceback (most recent call last):
  File "/Users/ed2050/Library/Python/3.8/lib/python/site-packages/anyio/streams/memory.py", line 98, in receive
    return self.receive_nowait()
  File "/Users/ed2050/Library/Python/3.8/lib/python/site-packages/anyio/streams/memory.py", line 93, in receive_nowait
    raise WouldBlock
anyio.WouldBlock

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/ed2050/Library/Python/3.8/lib/python/site-packages/starlette/middleware/base.py", line 78, in call_next
    message = await recv_stream.receive()
  File "/Users/ed2050/Library/Python/3.8/lib/python/site-packages/anyio/streams/memory.py", line 118, in receive
    raise EndOfStream
anyio.EndOfStream
```

# Code
Here's the code that triggered the error.  The problem is that mime.guess_type returns a tuple not a string.

Fortunately I had a log message just before the error that revealed the problem.  Usually it would be a lot harder to diagnose where the error occured.  This app has a few thousand lines over a dozen modules.

```py
import mimetypes as mime
from fastapi import app, Response

@app.get ('/retrieve/{ filename }')
def retrieve (filename) :

    contents = open (filename, 'r').read ()
    mtype = mime.guess_type (filename)  # <-- [0] here to fix

    log (f'returning { len (contents) } bytes with mime { mtype }')
    response = Response (content = contents, media_type = mtype)
    return response

# log output : returning 12462 bytes with mime ('image/jpeg', None) 

```

**Comment by falkoschindler** [comment]
Ok, I agree that good error messages are important and are part of what makes a library "nice" to use.

But in this specific case: How could NiceGUI intercept this exception? You're using `mime` from `mimetypes` and `Response` from `fastapi` before returning from a function decorated with FastAPI's `app.get`. NiceGUI isn't involved at all.

**Reply by ed2050** [reply]
Fair point.  Perhaps this particular example would be better directed at FastAPI.

That said, nicegui is built on fastapi.  It's not like a 3rd party library I chose.  There should be some role for reporting issues upstream to projects nicegui relies on.  Especially where this fastapi use case is [covered in nicegui docs](https://nicegui.io/documentation/section_pages_routing#api_responses).

**Reply by falkoschindler** [reply]
Sure, we do our best to prevent the user from running into unnecessarily confusing error messages. But in this case the code is completely independent of NiceGUI and I don't see how we could intercept this exception.

I'll close this discussion for now. But feel free to re-open it in case you have another example in mind.
